### PR TITLE
Improved PyTorch optimisation

### PR DESCRIPTION
**Pull request by yannmclatchie**
## Description

In order to improve the PyTorch optimisation loop, this PR will implement posterior draw thinning (similarly to the ``ndraws_pred`` argument in ``projpred`` optimisation) to improve optimisation speed. Further, we pivot from a family-specific definition of KL divergence towards the PyTorch pointwise loss function operating on log probabilities for simplicity and future maintenance ease.

## Checklist

- [x] Tests covering the new functionality have been added
- [x] Documentation has been updated OR the change is too minor to be documented
- [ ] Changes are listed in the `CHANGELOG.md` OR changes are insignificant

### CODE REVIEW

File: src/kulprit/data/submodel.py [hunk 1]
```diff
@@ -127,8 +127,8 @@ def create(
         """Convert some set of pytorch tensors into an ArviZ idata object.
 
         Args:
-            model (kulprit.ModelData): The restricted ModelData object whose
-                posterior to build
+            submodel_structure (kulprit.data.SubModelStructure): The restricted
+                model's structure object
```
Comment (tomicapretto) [inline src/kulprit/data/submodel.py:131]: Below we have 

```
        Returns:
            arviz.inferencedata: Restricted model idata object
```

it should be arviz.InferenceData

File: src/kulprit/data/submodel.py [hunk 2]
```diff
@@ -220,7 +220,7 @@ def posterior_to_points(self, posterior: dict) -> list:
             key: (
                 posterior[key].flatten()
                 if key in posterior.keys()
-                else np.zeros((self.data.structure.num_draws,))
+                else np.zeros((self.data.structure.num_thinned_draws,))
```
Comment (tomicapretto) [inline src/kulprit/data/submodel.py:223]: You can use `key in posterior` to check if a key is in a dictionary.

File: src/kulprit/data/submodel.py [hunk 3]
```diff
@@ -220,7 +220,7 @@ def posterior_to_points(self, posterior: dict) -> list:
...
             )
             for key in initial_point.keys()
```
Comment (tomicapretto) [inline src/kulprit/data/submodel.py:225]: Same comment about the usage of `.keys()`

File: src/kulprit/projection/losses.py [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+    This class computes some KL divergence loss for observations seen from the
+    the reference model variate's family. The KL divergence is the originally
+    motivated loss function by Goutis and Robert (1998).
+    """
+
+    def __init__(self) -> None:
+        """Loss module constructor."""
+        super(KullbackLeiblerLoss, self).__init__()
+
+        self.loss = nn.KLDivLoss(reduction="batchmean", log_target=True)
```
Comment (tomicapretto) [inline src/kulprit/projection/losses.py:27]: Do you think this detail, `reduction="batchmean"`, is something that could go into the docstring? I'm not sure if it's important. But if you consider it's not obvious, it may be good to clarify that in the docstring. If it's a minor implementation detail, feel free to leave it as it is.

File: src/kulprit/projection/projector.py [hunk 5]
```diff
@@ -15,7 +15,7 @@ def __init__(
         self,
         data: ModelData,
         path: Optional[SearchPath] = None,
-        num_iters: Optional[int] = 200,
+        num_iters: Optional[int] = 400,
```
Comment (tomicapretto) [inline src/kulprit/projection/projector.py:18]: The docstring does not document all the arguments

File: src/kulprit/projection/losses.py (hunk 4 above)
Comment (yannmclatchie) [reply src/kulprit/projection/losses.py]: I hadn't considered putting this in the docstring, but it would actually be valuable to document it. In essence the reason for this setting is the way PyTorch compute the point-wise KL divergence, and so it would be nice to mention this for sure.
